### PR TITLE
Port most of the existing Kafka Source tests to rekt framework

### PR DESCRIPTION
**Pull request by lionelvillard**
Fixes #

## Proposed Changes

-  Add KafkaSource, KafkaTopic and Kafkacat resources.
-  Add KafkaSource and KafkaTopic readiness features
- Add KafkaSource data plane feature
- Port all KafkaSource tests except the ones updating the source.
- Run rekt tests only in the mt-source prow job (for now). 

**Release Note**

```release-note

```

**Docs**

To be used by https://github.com/knative-sandbox/eventing-autoscaler-keda/pull/109

### CODE REVIEW

**Comment by lionelvillard** [comment]
@slinkydeveloper Can you take a look at this, since you know rekt ?

File: test/rekt/features/kafkasource/data_plane.go
Comment (n3wscott) [inline test/rekt/features/kafkasource/data_plane.go:49]: I would not put this at the assert timing level, should be a `f.Requirement` as it is needed for the test, but not the test.

File: test/rekt/features/kafkasource/data_plane.go
Comment (n3wscott) [inline test/rekt/features/kafkasource/data_plane.go:34]: part of this struggle is to think about what the invariant of the test is and what you want the test to be doing.

I would think about reframing the test more like: "Given this kafka source (it is made before the feature is invoked), and this event replayer instance (could be passed as a StepFn from `kafkacat.Install(name, kcopts...)` and then added to or called in a `Must`, did all the events get seen?

Comment (n3wscott) [reply test/rekt/features/kafkasource/data_plane.go:34]: So I might think to change the signature from 

```
func DataPlaneDelivery(name string, ksopts []kafkasource.CfgFn, kcopts []kafkacat.CfgFn, matchers cetest.EventMatcher) *feature.Feature {
```

to

```
func DataPlaneDelivery(name string, produceEvents feature.StepFn, matchers cetest.EventMatcher) *feature.Feature {
```

- No `ksopts []kafkasource.CfgFn`, because the source already exists.
- No `kcopts []kafkacat.CfgFn` because it is passed as an event production step function.

File: test/rekt/features/kafkasource/data_plane.go
Comment (lionelvillard) [reply test/rekt/features/kafkasource/data_plane.go:49]: Yes make sense!

File: test/rekt/features/kafkasource/data_plane.go
Comment (lionelvillard) [reply test/rekt/features/kafkasource/data_plane.go:34]: @n3wscott Did you have time to give more thoughts on this?

File: test/rekt/resources/kafkatopic/topic.go
Comment (n3wscott) [inline test/rekt/resources/kafkatopic/topic.go:37]: nit before we get too far in copy/paste, `GVR`

Comment (lionelvillard) [reply test/rekt/resources/kafkatopic/topic.go:37]: fixed.

File: test/rekt/features/kafkasource/data_plane.go
Comment (slinkydeveloper) [inline test/rekt/features/kafkasource/data_plane.go:91]: Sending the event IMO should happen in the `Setup` phase, following the Given-When-Then pattern (where setup is both Given and When) https://github.com/knative-sandbox/reconciler-test/issues/106

File: test/rekt/features/kafkasource/readiness.go
Comment (slinkydeveloper) [inline test/rekt/features/kafkasource/readiness.go:41]: Should this be an assert?

File: test/rekt/features/kafkatopic/readiness.go
Comment (lionelvillard) [reply test/rekt/features/kafkatopic/readiness.go:31]: @n3wscott suggested it to be a Requirement. 

Anyhow IMO there is not a clear distinction between Requirement vs Assert and maybe it should be removed, unless it can be clearly be explained.

File: test/rekt/features/kafkasource/data_plane.go
Comment (lionelvillard) [reply test/rekt/features/kafkasource/data_plane.go:67]: it feels more like an assertion to me.

File: test/rekt/features/kafkasource/data_plane.go
Comment (lionelvillard) [reply test/rekt/features/kafkasource/data_plane.go:91]: not sure how this can be done because there is no order guarantee between steps in the same phase.

File: test/rekt/kafka_source_test.go
Comment (lionelvillard) [reply test/rekt/kafka_source_test.go:247]: yes good point, I meant to look at featureset. thanks!

File: test/rekt/features/kafkasource/data_plane.go
Comment (slinkydeveloper) [reply test/rekt/features/kafkasource/data_plane.go:91]: @lionelvillard there is ordering guarantee. Phases are executed sequentially

Comment (lionelvillard) [reply test/rekt/features/kafkasource/data_plane.go:91]: yes but not within a single phase

File: test/rekt/kafka_source_test.go
Comment (lionelvillard) [reply test/rekt/kafka_source_test.go:247]: I remove the wrapping function. My goal here is to give an unique name to each test, so I added the test name to the feature name.

**Comment by matzew** [comment]
/assign
will review tomorrow

**Comment by lionelvillard** [comment]
@matzew I've made the changes I wanted to make. Please TAL, thanks!

**Comment by lionelvillard** [comment]
gentle ping. @matzew

**Comment by lionelvillard** [comment]
@matzew Did you have a chance to look at this?

File: test/e2e-common.sh
Comment (matzew) [inline test/e2e-common.sh:523]: does this TLS?

Comment (lionelvillard) [reply test/e2e-common.sh:523]: yes

**Comment by lionelvillard** [comment]
@knative-sandbox/eventing-kafka-approvers @knative-sandbox/eventing-wg-leads Is there anything I can do to help getting this PR moving forward (besides rebasing)? I need to make a decision on whether to host our e2e tests here or somewhere else. If I get no reply by tomorrow, I'll close this PR.

File: test/rekt/README.md
Comment (matzew) [inline test/rekt/README.md:15]: I think we need a bit more here, but we can add that in a later PR

e.g. a proper strimzi needs to be around, like the setup does for the CI on the repo.

I also got:

```
2021/05/03 17:27:55 Published docker.io/matzew/eventshub@sha256:4a6dd402d1cf0592e89825ed733cb4262a558cef104d68fcd36955b9d76f12d8
    kafka_source_test.go:54: could not copy secret(strimzi-tls-secret): secrets "strimzi-tls-secret" not found
--- FAIL: TestKafkaSource (16.95s)

```

due to missing setup instructions - but that's NOT a blocker for this PR

File: test/rekt/resources/kafkacat/configmap.yaml
Comment (matzew) [inline test/rekt/resources/kafkacat/configmap.yaml:21]: how/where is the configmap used?

File: test/rekt/resources/kafkacat/kafkacat_test.go
Comment (matzew) [inline test/rekt/resources/kafkacat/kafkacat_test.go:74]: is this added b/c it's needed? or more to visualize the actual output?

File: test/rekt/resources/kafkacat/kafkacat_test.go
Comment (matzew) [inline test/rekt/resources/kafkacat/kafkacat_test.go:84]: play load? :smile:

File: test/rekt/resources/kafkacat/kafkacat_test.go
Comment (lionelvillard) [reply test/rekt/resources/kafkacat/kafkacat_test.go:74]: yeah it's needed. See this: https://blog.golang.org/examples

File: test/rekt/resources/kafkacat/configmap.yaml
Comment (lionelvillard) [reply test/rekt/resources/kafkacat/configmap.yaml:21]: when Kafkacat is installed (https://github.com/knative-sandbox/eventing-kafka/pull/488/files#diff-c08ede98b036f16224e9ab7f7a22b8ac1e29498c62254c35df90cf854ed63d56R39), both the deployment and the configmap are created.

File: test/rekt/README.md
Comment (lionelvillard) [reply test/rekt/README.md:15]: I'll open an issue to keep track of it

File: test/rekt/features/kafkasource/data_plane.go
Comment (mgencur) [reply test/rekt/features/kafkasource/data_plane.go:296]: The same source of confusion as in the other places: I also expressed confusion about using Requirement in knative-sandbox/reconciler-test#106 (comment) . This single check we're commenting on could easily be wrapped/verified in the Setup step a few lines above. I believe this will continue to be a source of confusion until the docs precisely specify its purpose.

File: test/rekt/features/kafkasource/readiness.go
Comment (mgencur) [inline test/rekt/features/kafkasource/readiness.go:28]: Is this function used anywhere? I don't see it in this PR.

File: test/rekt/resources/kafkacat/kafkacat_test.go
Comment (mgencur) [inline test/rekt/resources/kafkacat/kafkacat_test.go:25]: Nit: Why not `ExampleMin` and `ExampleFull` ? AFAIK, that's the "case" we use.

Comment (mgencur) [reply test/rekt/resources/kafkacat/kafkacat_test.go:25]: More important: What's the purpose of these functions? It's not a test function but it's in a file with `_test.go` suffix. I don't see this function being called anywhere. Is it some temporary stuff?

**Review by mgencur — CHANGES_REQUESTED** [review]
Hi @lionelvillard , overall, I think it goes in the direction that was set by REKT and I think it looks pretty good. I'm just confused by a few things which I commented on - such as the usage of Requirement vs. Assert vs. Setup. Also, some files seem to be unused. Overall, it's a good piece of work :-)

File: pkg/client/injection/reconciler/bindings/v1alpha1/kafkabinding/state.go
Comment (lionelvillard) [reply pkg/client/injection/reconciler/bindings/v1alpha1/kafkabinding/state.go:32]: hmm the update-deps GitHub action is not running anymore. This is generated code. Please ignore.

File: pkg/client/injection/reconciler/bindings/v1alpha1/kafkabinding/state.go
Comment (lionelvillard) [reply pkg/client/injection/reconciler/bindings/v1alpha1/kafkabinding/state.go:36]: same here, generated code.

File: pkg/client/injection/reconciler/bindings/v1beta1/kafkabinding/state.go
Comment (lionelvillard) [reply pkg/client/injection/reconciler/bindings/v1beta1/kafkabinding/state.go:36]: same

File: pkg/client/injection/reconciler/bindings/v1beta1/kafkabinding/state.go
Comment (lionelvillard) [reply pkg/client/injection/reconciler/bindings/v1beta1/kafkabinding/state.go:40]: same

File: test/rekt/features/kafkasource/readiness.go
Comment (lionelvillard) [reply test/rekt/features/kafkasource/readiness.go:28]: you're right, it's not used yet. It should probably be added to the smoke tests.

File: test/rekt/features/kafkatopic/readiness.go
Comment (lionelvillard) [reply test/rekt/features/kafkatopic/readiness.go:31]: I agree. Let's keep `Requirement` for now and we may revisit later on. Ok?

File: test/rekt/resources/kafkacat/kafkacat_test.go
Comment (lionelvillard) [reply test/rekt/resources/kafkacat/kafkacat_test.go:25]: This is correct. See https://blog.golang.org/examples#TOC_4. 

I actually really like this tests since they help developing/debugging the YAML templates. `go test ./...` runs these functions.

Comment (mgencur) [reply test/rekt/resources/kafkacat/kafkacat_test.go:25]: Oh. Thanks. Didn't know that.

File: test/rekt/features/kafkasource/readiness.go
Comment (mgencur) [reply test/rekt/features/kafkasource/readiness.go:28]: Would you like to add this later or in this PR? Just to make it clear and be able to move ahead.

File: test/rekt/features/kafkatopic/readiness.go
Comment (mgencur) [reply test/rekt/features/kafkatopic/readiness.go:31]: OK

File: test/rekt/features/kafkasource/readiness.go
Comment (lionelvillard) [reply test/rekt/features/kafkasource/readiness.go:28]: I can add this later.

**Comment by lionelvillard** [comment]
/hold

I might be using an old API, thus the go.mod diff.

**Comment by lionelvillard** [comment]
nvm, it was a bad rebase. fixed.

/unhold

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
